src/libsync/clientstatusreporting.cpp

```diff
@@ -0,0 +1,440 @@
+{
+Q_LOGGING_CATEGORY(lcClientStatusReporting, "nextcloud.sync.clientstatusreporting", QtInfoMsg)
+
+ClientStatusReporting::ClientStatusReporting(Account *account, QObject *parent)
+    : QObject(parent)
+    , _account(account)
+{
+    init();
+}
+
+ClientStatusReporting::~ClientStatusReporting()
+{
+    if (_database.isOpen()) {
+        _database.close();
+    }
+}
+
+void ClientStatusReporting::init()
+{
+    Q_ASSERT(!_isInitialized);
+    if (_isInitialized) {
+        qCDebug(lcClientStatusReporting) << "Double call to init";
+        return;
+    }
+
+    for (int i = 0; i < ClientStatusReporting::Status::Count; ++i) {
+        const auto statusString = statusStringFromNumber(static_cast<Status>(i));
+        _statusNamesAndHashes[i] = {statusString, c_jhash64((uint8_t *)statusString.data(), statusString.size(), 0)};
+    }
+
+    const auto dbPath = makeDbPath();
+    _database = QSqlDatabase::addDatabase(QStringLiteral("QSQLITE"));
+    _database.setDatabaseName(dbPath);
+
+    if (!_database.open()) {
+        qCDebug(lcClientStatusReporting) << "Could not setup client reporting, database connection error.";
+        return;
+    }
+
+    QSqlQuery query;
+    const auto prepareResult = query.prepare(QStringLiteral(
+        "CREATE TABLE IF NOT EXISTS clientstatusreporting("
+        "name VARCHAR(4096) PRIMARY KEY,"
+        "status INTEGER(8),"
+        "count INTEGER,"
+        "lastOccurrence INTEGER(8))"));
+    if (!prepareResult || !query.exec()) {
+        qCDebug(lcClientStatusReporting) << "Could not setup client clientstatusreporting table:" << query.lastError().text();
+        return;
+    }
+
+    if (!query.prepare(QStringLiteral("CREATE INDEX IF NOT EXISTS name ON clientstatusreporting(name);")) || !query.exec()) {
+        qCDebug(lcClientStatusReporting) << "Could not create index on clientstatusreporting table:" << query.lastError().text();
+        return;
+    }
+
+    if (!query.prepare(QStringLiteral("CREATE TABLE IF NOT EXISTS keyvalue(key VARCHAR(4096), value VARCHAR(4096), PRIMARY KEY(key))")) || !query.exec()) {
+        qCDebug(lcClientStatusReporting) << "Could not setup client keyvalue table:" << query.lastError().text();
+        return;
+    }
+
+    // prevent issues in case enum gets changed in future, hash its value and clean the db in case there was a change
+    QByteArray statusNamesContatenated;
+    for (int i = 0; i < ClientStatusReporting::Status::Count; ++i) {
+        statusNamesContatenated += statusStringFromNumber(static_cast<Status>(i));
+    }
+    statusNamesContatenated += QByteArray::number(ClientStatusReporting::Status::Count);
+    const auto statusNamesHashCurrent = QCryptographicHash::hash(statusNamesContatenated, QCryptographicHash::Md5).toHex();
+    const auto statusNamesHashFromDb = getStatusNamesHash();
+
+    if (statusNamesHashCurrent != statusNamesHashFromDb) {
+        deleteClientStatusReportingRecords();
+        setStatusNamesHash(statusNamesHashCurrent);
+    }
```

Comment:
please make it a method with a self explaining name
that may even remove the need for the comment :smile: